### PR TITLE
Rover Sailboat Support

### PR DESCRIPTION
**Pull request by IamPete1**
On to number 3, continuation of https://github.com/ArduPilot/ardupilot/pull/9322

### CODE REVIEW

**Comment by rmackay9** [comment]
@IamPete1,

I've put a [fix for the issue in my iampete1-sailboat2 branch](https://github.com/rmackay9/rmackay9-ardupilot/commits/iampete1-sailboat2).  The issue was the airspeed prefix was one character too long.  By the way, if you suspect this kind of problem, if you start Rover in the simulator, the error is often written in a console window.

**Comment by IamPete1** [comment]
Thanks for the fix, i'm coming to realize if it complies but doesn't run its usually to do with parameters being in the wrong place or to too long. Hopefully I can get the new sensor running today.

File: APMrover2/Log.cpp [hunk 1]
```diff
@@ -154,6 +154,41 @@ void Rover::Log_Write_Proximity()
...
+    // only log sail if present
+    if (!g2.motors.has_sail()) {
+        return;
+    }
+
+    // get wind direction
+    float wind_dir_abs = 0.0f;
+    float wind_dir_rel = 0.0f;
+    float true_wind_speed = 0.0f;
+    if (rover.g2.windvane.enabled()){
```
Comment (khancyr) [inline APMrover2/Log.cpp]: if (rover.g2.windvane.enabled()) {

File: APMrover2/Log.cpp [hunk 2]
```diff
@@ -154,6 +154,41 @@ void Rover::Log_Write_Proximity()
...
+        wind_dir_abs = degrees(g2.windvane.get_absolute_wind_direction_rad());
+        wind_dir_rel = degrees(g2.windvane.get_apparent_wind_direction_rad());
+        true_wind_speed = g2.windvane.get_true_wind_speed();
+    } 
+    float aparent_wind_speed = 0.0f;
+    if (rover.g2.airspeed.enabled()){
```
Comment (khancyr) [inline APMrover2/Log.cpp]: if (rover.g2.airspeed.enabled()) {

File: APMrover2/Log.cpp [hunk 3]
```diff
@@ -154,6 +154,41 @@ void Rover::Log_Write_Proximity()
...
+        aparent_wind_speed = rover.g2.airspeed.get_airspeed();
+    }
+    float VMG = 0.0f;
+    if (rover._sailboat_indirect_route){
```
Comment (khancyr) [inline APMrover2/Log.cpp]: if (rover._sailboat_indirect_route) {

File: APMrover2/Log.cpp [hunk 4]
```diff
@@ -154,6 +154,41 @@ void Rover::Log_Write_Proximity()
...
+    if (rover.g2.windvane.enabled()){
+        wind_dir_abs = degrees(g2.windvane.get_absolute_wind_direction_rad());
+        wind_dir_rel = degrees(g2.windvane.get_apparent_wind_direction_rad());
+        true_wind_speed = g2.windvane.get_true_wind_speed();
+    } 
+    float aparent_wind_speed = 0.0f;
+    if (rover.g2.airspeed.enabled()){
+        aparent_wind_speed = rover.g2.airspeed.get_airspeed();
+    }
+    float VMG = 0.0f;
```
Comment (khancyr) [inline APMrover2/Log.cpp]: vmg, majuscule is for constant

File: APMrover2/Log.cpp [hunk 5]
```diff
@@ -154,6 +154,41 @@ void Rover::Log_Write_Proximity()
...
+    if (rover._sailboat_indirect_route){
+        VMG = rover._sailboat_velocity_made_good;
+    }
+    DataFlash.Log_Write("SAIL", "TimeUS,WindDirAbs,WindDirRel,TrueWindSpd,AparentWindSpd,SailOut,VMG",
+                        "shh%", "F000", "Qfff", // not sure if i need to change this line or not??
```
Comment (khancyr) [inline APMrover2/Log.cpp]: see : http://ardupilot.org/dev/docs/code-overview-adding-a-new-log-message.html#the-easy-way
first is units (https://github.com/ArduPilot/ardupilot/blob/master/libraries/DataFlash/LogStructure.h#L81), second multiplier, last format.
So it should be
"s--n%n", "F------", "Qffffff",
Standing for 
second, unknown, unknown, m/s, percent, m/s

File: APMrover2/mode.cpp [hunk 6]
```diff
@@ -448,6 +464,12 @@ void Mode::calc_steering_from_lateral_acceleration(float lat_accel, bool reverse
 // calculate steering output to drive towards desired heading
 void Mode::calc_steering_to_heading(float desired_heading_cd, float rate_max, bool reversed)
 {
+    // if we cant sail at desired heading caculate new heading to sailing on, also update maximum rate
+    if (rover.sailboat_update_indirect_route(desired_heading_cd)){
```
Comment (khancyr) [inline APMrover2/mode.cpp]: if (rover.sailboat_update_indirect_route(desired_heading_cd)) {

File: APMrover2/mode_acro.cpp [hunk 7]
```diff
@@ -11,23 +11,39 @@ void ModeAcro::update()
...
     }
 
-    // convert pilot steering input to desired turn rate in radians/sec
-    const float target_turn_rate = (desired_steering / 4500.0f) * radians(g2.acro_turn_rate);
+    float steering_out = 0.0f;
 
-    // run steering turn rate controller and throttle controller
-    const float steering_out = attitude_control.get_steering_out_rate(target_turn_rate,
+    // Check if we should to a tack due to aux switch
+    if (g2.motors.has_sail() && (rover._sailboat_tack || rover._sailboat_tacking)){
```
Comment (khancyr) [inline APMrover2/mode_acro.cpp]: if (g2.motors.has_sail() && (rover._sailboat_tack || rover._sailboat_tacking)) {

File: APMrover2/mode_hold.cpp [hunk 8]
```diff
@@ -4,13 +4,39 @@
...
         rover.balancebot_pitch_control(throttle);
     }
 
+    // if we have a sail we should let it out and steer into the wind
+    if (g2.motors.has_sail()){
+        // relax mainsail
+        rover.sailboat_set_mainsail(100);
+        
+        // Only steer into wind if armed and a sterring angle is deffined
+        if(hal.util->get_soft_armed() && !is_negative(g2.sailboat_hold_angle)){
```
Comment (khancyr) [inline APMrover2/mode_hold.cpp]: if (hal.util->get_soft_armed() && !is_negative(g2.sailboat_hold_angle)) {

File: APMrover2/mode_hold.cpp [hunk 9]
```diff
@@ -4,13 +4,39 @@
...
     float throttle = 0.0f;
+    float steering_out = 0.0f; 
 
     // if vehicle is balance bot, calculate actual throttle required for balancing
     if (rover.is_balancebot()) {
         rover.balancebot_pitch_control(throttle);
     }
 
+    // if we have a sail we should let it out and steer into the wind
+    if (g2.motors.has_sail()){
```
Comment (khancyr) [inline APMrover2/mode_hold.cpp]: if (g2.motors.has_sail()) {

File: APMrover2/mode_hold.cpp [hunk 10]
```diff
@@ -4,13 +4,39 @@
...
+        // relax mainsail
+        rover.sailboat_set_mainsail(100);
+        
+        // Only steer into wind if armed and a sterring angle is deffined
+        if(hal.util->get_soft_armed() && !is_negative(g2.sailboat_hold_angle)){
+            if(is_zero(g2.sailboat_hold_angle)){
```
Comment (khancyr) [inline APMrover2/mode_hold.cpp]: if (is_zero(g2.sailboat_hold_angle)) {

File: APMrover2/mode_hold.cpp [hunk 11]
```diff
@@ -4,13 +4,39 @@
...
+                // call heading controller
+                steering_out = attitude_control.get_steering_out_heading(g2.windvane.get_absolute_wind_direction_rad(),
+                                                                        g2.pivot_turn_rate,
+                                                                        g2.motors.limit.steer_left,
+                                                                        g2.motors.limit.steer_right,
+                                                                        rover.G_Dt);
+            } else {
+                // if hold angle is not zero use sailboat heading controler, this allows aux switch and rudder input tacking, pass the wind direciton as the defualt heading in case of error
+                steering_out = attitude_control.get_steering_out_heading(radians(rover.sailboat_calc_heading(g2.windvane.get_absolute_wind_direction_rad()) / 100.0f),
```
Comment (khancyr) [inline APMrover2/mode_hold.cpp]: * 0.01f

File: APMrover2/radio.cpp [hunk 12]
```diff
@@ -79,24 +79,27 @@ void Rover::rudder_arm_disarm_check()
...
-        if (channel_steer->get_control_in() < -4000) {
-            const uint32_t now = millis();
-
-            if (rudder_arm_timer == 0 ||
-                now - rudder_arm_timer < ARM_DELAY_MS) {
-                if (rudder_arm_timer == 0) {
-                    rudder_arm_timer = now;
+    } else if (!g2.motors.active() && g2.rudder_arming_value != AP_Arming_Rover::ARMING_RUDDER_ARMONLY) {
+        // Dont allow  disarm if in manual and rudder arming parameter is 3, for sailboats
+        if (!(g2.rudder_arming_value == AP_Arming_Rover::ARMING_RUDDER_NO_DISARM_MANUAL && rover.control_mode == &rover.mode_manual)){
```
Comment (khancyr) [inline APMrover2/radio.cpp]: )) {

File: APMrover2/sailboat.cpp [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+    g2.motors.set_mainsail(mainsail);
+}
+
+// Should we take a indirect navigaion route, either to go upwind or in the future for speed 
+bool Rover::sailboat_update_indirect_route(float desired_heading)
+{
+    if (!g2.motors.has_sail()) {
+        return false;
+    }
+    desired_heading = radians(desired_heading / 100.0f);
```
Comment (khancyr) [inline APMrover2/sailboat.cpp]: * 0.01f

File: APMrover2/sailboat.cpp [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+    _sailboat_indirect_route = false;
+    
+    // Check if desired heading is in the no go zone, if it is we can't go direct
+    // add 10 deg padding to try and avoid constant switching between methods, maybe add a 'dead zone'?
+    if (fabsf(wrap_PI((g2.windvane.get_absolute_wind_direction_rad() - desired_heading))) <= radians(g2.sail_no_go + 10.0f)){
```
Comment (khancyr) [inline APMrover2/sailboat.cpp]: )) {

File: APMrover2/sailboat.cpp [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+}    
+
+// If we can't sail on the desired heading then we should pick the best heading that we can sail on 
+float Rover::sailboat_calc_heading(float desired_heading)
+{
+    if (!g2.motors.has_sail()) {
+        return desired_heading;
+    }
+    
+    desired_heading = radians(desired_heading / 100.0f);
```
Comment (khancyr) [inline APMrover2/sailboat.cpp]: * 0.01f

File: APMrover2/sailboat.cpp [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+        left_no_go_heading = wrap_2PI(g2.windvane.get_absolute_wind_direction_rad() + radians(g2.sailboat_hold_angle));
+        right_no_go_heading = wrap_2PI(g2.windvane.get_absolute_wind_direction_rad() - radians(g2.sailboat_hold_angle));
+    } else {
+        // Use upwind tacking angles
+        left_no_go_heading = wrap_2PI(g2.windvane.get_absolute_wind_direction_rad() + radians(g2.sail_no_go));
+        right_no_go_heading = wrap_2PI(g2.windvane.get_absolute_wind_direction_rad() - radians(g2.sail_no_go));
+    }
+
+    // Caculate what tack we are on if it has been too long since we knew
+    if (_sailboat_current_tack ==  _tack::Unknown || (AP_HAL::millis() - _sailboat_heading_last_run) > 1000){
```
Comment (khancyr) [inline APMrover2/sailboat.cpp]: ) {

File: APMrover2/sailboat.cpp [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+        if (g2.windvane.get_apparent_wind_direction_rad() < 0){
```
Comment (khancyr) [inline APMrover2/sailboat.cpp]: if (is_negatif(g2.windvane.get_apparent_wind_direction_rad())) {

File: APMrover2/sailboat.cpp [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+            _sailboat_current_tack = _tack::Port;
+        } else {
+            _sailboat_current_tack = _tack::STBD;
+        }   
+    }
+    _sailboat_heading_last_run = AP_HAL::millis();
+    
+    // Allow force tack from rudder input 
+    float steering_in = rover.channel_steer->norm_input();
+    if (fabsf(steering_in) > 0.9f && !_sailboat_tack && !_sailboat_tacking){
```
Comment (khancyr) [inline APMrover2/sailboat.cpp]: where does 0.9f come ? lets make define or static const with a comment !

File: APMrover2/sailboat.cpp [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+       
+        switch (_sailboat_current_tack){
```
Comment (khancyr) [inline APMrover2/sailboat.cpp]: arf problem with ){ everywhere ...

File: APMrover2/sailboat.cpp [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+                break;
+            }  
+        }        
+    }
+    
+    // Maximum cross track error before tack, this efectively defines a 'corridor' of width 2*sailboat_auto_xtrack_tack that the boat will stay within, disable if tacking or in hold mode
+    if (fabsf(rover.nav_controller->crosstrack_error()) >= g2.sailboat_auto_xtrack_tack && !is_zero(g2.sailboat_auto_xtrack_tack) && !_sailboat_tack && !_sailboat_tacking && rover.control_mode != &rover.mode_hold){
+        // Make sure the new tack will reduce the cross track error        
+        // If were on starbard tack we a travling towards the left hand boundary
+        if (rover.nav_controller->crosstrack_error() > 0 && _sailboat_current_tack == _tack::STBD){
```
Comment (khancyr) [inline APMrover2/sailboat.cpp]: is_positif ?

File: APMrover2/sailboat.cpp [hunk 21]
```diff
@@ -0,0 +1,268 @@
...
+            _sailboat_tack = true;
+        }
+        // If were on port tack we a travling towards the right hand boundary
+        if (rover.nav_controller->crosstrack_error() < 0 && _sailboat_current_tack == _tack::Port){
```
Comment (khancyr) [inline APMrover2/sailboat.cpp]: is_negative ?

File: libraries/APM_Control/AR_AttitudeControl.cpp [hunk 22]
```diff
@@ -554,6 +598,58 @@ float AR_AttitudeControl::get_desired_pitch() const
...
+    // heel error is given as input to PID contoller
+    _sailboat_heel_pid.set_input_filter_all(heel_error);
+
+    // get feed-forward
+    const float ff = _sailboat_heel_pid.get_ff(desired_heel);
+
+    // get p
+    float p = _sailboat_heel_pid.get_p();
+    // constrain p to only be positive 
+    if(!is_positive(p)){
```
Comment (khancyr) [inline libraries/APM_Control/AR_AttitudeControl.cpp]: if (!is_positive(p)) {

File: libraries/APM_Control/AR_AttitudeControl.cpp [hunk 23]
```diff
@@ -554,6 +598,58 @@ float AR_AttitudeControl::get_desired_pitch() const
...
+        p = 0.0f; 
+    }
+    
+    // get i
+    float i = _sailboat_heel_pid.get_integrator();
+    // constrain i to only be positive, reset integrator if negative
+    if(!is_positive(i)){
```
Comment (khancyr) [inline libraries/APM_Control/AR_AttitudeControl.cpp]: if (!is_positive(i)) {

File: libraries/AP_WindVane/AP_WindVane.h [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+    // pin for reading analog voltage
+    AP_HAL::AnalogSource *windvane_analog_source;
+
+    // pointer to airspeed object, if available
+    AP_Airspeed* _airspeed;
+};
+
+namespace AP {
+    AP_WindVane *windvane();
+};
```
Comment (khancyr) [inline libraries/AP_WindVane/AP_WindVane.h]: miss empty line at the end (it helps diff tools)

**Comment by IamPete1** [comment]
@khancyr thanks for the review, I think I got everything

**Comment by IamPete1** [comment]
Its failing on px4-v3 using make, looks like something to do with adding the airspeed library, its fine for everything else.

File: APMrover2/GCS_Mavlink.cpp [hunk 25]
```diff
@@ -264,6 +264,33 @@ void Rover::send_fence_status(mavlink_channel_t chan)
     fence_send_mavlink_status(chan);
 }
 
+void Rover::send_wind(mavlink_channel_t chan)
```
Comment (peterbarker) [inline APMrover2/GCS_Mavlink.cpp]: This should be in `GCS_MAVLINK_Rover`

Comment (IamPete1) [reply APMrover2/GCS_Mavlink.cpp]: you can blame @rmackay9 for this one, I have had a quick look, all the other send_... functions are under Rover for nav controller, servos out, rangefinder, pids ect. I am happy to do whichever, thanks for having a look.

**Comment by IamPete1** [comment]
I have rebased on master I think the code is complete enough to start looking to merge. I fairly happy it all works, only untested thing is automatic tacking on geo-fence. Apart from that I think everything works as intended. I guess the best thing to do is to merge and do some documentation then hopefully a few people will have a go and then we can fix things as we find them.

I guess I need to fix whatever is causing it to fail on make and quad plane SITL. Not to sure whats wrong in either case. 

I will have a final check through the code update the some of the parameter defaults and spell check the comments. I guess then squash the commits abit and its ready to go?

**Comment by rmackay9** [comment]
@IamPete1, great stuff.  Yes, I think a final check from your side and squashing some commits together would be great.  then the review wolves (myself included) will have at it.

I'm happy to help figure out why make and quadplane autotesting is failing..

**Comment by IamPete1** [comment]
Have squashed the commits and done a final check. Went and tested the it on the water this afternoon all worked well, vid to follow on forum. I think its now ready for review. The only thing i'm not sure about is whether or not is worth adding airspeed sensor logging. The only reason the airspeed sensor would be enabled would be to use it as a windvane in which case the speed you be logged in the sailing log. I guess it wouldn't do any harm to have air speed logging too. 

@rmackay9 It seems to have passed the autotest this time, I think its something to do with the auto test rather than the code. If you could have a look at the make issue that would be great!, Thanks

**Comment by rmackay9** [comment]
@IamPete1, great stuff.  I'm going through it now and I'll at least fix-up all the obvious stuff that I know other reviews will point out.  I'll post back here once I have something.. that'll probably be tomorrow.

**Comment by IamPete1** [comment]
@rmackay9  Thank You, hope its not to much off a mess. It looks like you managed to 'un-commit' some stuff in order to recommit it in a more sensible order? I had a look how to do that couldn't work it out. Hopefully the second half of the commits are not such a mess once you suggested I try and spit them up in to different library's ect. Hope my spelling is not too bad, not one of my strong points! 

Thanks Again

**Comment by rmackay9** [comment]
@IamPete1,
yes, i squashed and then extracted some bits (like adding AP_Airspeed support) into new commits.  I use SourceTree (Windows only I think) and it includes the ability to stage just a few lines of a file that's been modified.

I've added a few more commits to my repo's iampete1-sailboat5 repo mostly in the AP_WindVane lib.  The changes i've done are mostly non-functional (i.e. renaming variables).  The only changes of note are:

- allow calibration to run at any time (not just after startup).
- removed the checks around calling the filter's set_cutoff_frequency().  Removing these checks does lead to a few more multiplies happening but the CPU cost is really tiny so I thought the optimisation was probably not worth the complexity.  There was actually a small typo in the speed-filter's check so it was being called on each loop anyway.

A few other comments/ideas:

- maybe we shouldn't call read_PWM_bearing directly from get_absolute_wind_direction_rad, instead we should read the pwm value as part of the update() call?  This would make it more consistent with how we read from the analog pin.  It will also allow us to make get_absolute_wind_direction_rad const.  Not a big deal of course.
- I wonder if the windspeed sensor code should be part of AP_Airspeed.  I know we discussed this before and I haven't looked into how complex AP_Airspeed is and how difficult the fit would be.
- Log_Write_Sail should be logged at 10hz I think instead of at the main loop rate

I haven't looked too much at the vehicle code yet but I'll do that next.

It's all pretty nice..

**Comment by rmackay9** [comment]
@iampete1,
How would you feel about removing the ability to tack from Hold and Acro modes?  I think for Hold mode, we normally do not accept user input so I think this makes sense at least.
My motivation is just to try and keep things simple..

**Comment by IamPete1** [comment]
@rmackay9 Thanks for all this extra work, 

- putting read_PWM_bearing in the update() makes sence 
- I did start out doing the sensor in the airspeed library but it was ending up that i was having to add so much stuff it would make sense just to put in the the windvane lib. Because the sensor reads the airspeed direct, the airspeed lib is expecting a pressure. The auto offset stuff would need a few changes to work and it needs a second analog pin. I didn't want to mess up the airspeed lib by adding a special case for this one sensor at the end of each function. I think it makes sense for sensors that would only be used on sailboats like this one.
- Yeah log_Write_Sail should be at 10hz, i just put it in quickly without thinking too much 
- For acro and hold tacking, I think acro mode tacking is defiantly useful, it give you a chance to test the tacking rate (pivot rate param) without being in auto, also you can just set the boat at a nice angle to the wind and tack upwind  just using the just the switch.  Hold mode tacking is more of a edge case that will only be useful for a small number of boats where you want to stop but not get stuck head to wind. For some boats in some conditions it is possible to get stuck head to wind and not be able to get out by any movement of rudder. In this case you want to keep minimal amount of speed sort of like https://en.wikipedia.org/wiki/Heaving_to if your moving slowly then the tack function is useful.  If your trying to keep it simple for the user we could have a parameter for tuning one or both off and default to both off. If your trying to keep the code simple that wont help at all.

Thanks,

**Comment by rmackay9** [comment]
@IamPete1,

ok, i guess we can think about this acro and hold issue a bit more.  The problem I see is that sailboat_calc_heading doesn't fit in with the architecture because it checks the vehicle's flight mode and it checks user input.

the processing of user input really should be done near the top of a mode_xxx.cpp file and instead of direct checks of the flight mode, the mode should have new attributes or methods added.

more fundamentally though, I don't really think hold should be doing complex operations like tacking.  Hold mode is currently a very simple mode that pretty much just turns off user input.  It's almost the same as disarming the vehicle.  I feel it's getting turned into something closer to Loiter..

**Comment by IamPete1** [comment]
@rmackay9 

I see what you mean, hold should defiantly do some steering, we could leave it as just steering straight into the wind. When i was first testing and tried the existing hold it was OK if you were going up wind, but if you happened to be going down wind when you engaged hold it would just carry on straight with the sails out for maximum power and the rudder locked.

Then we could save the 'hold angles' and hold tacking for a future PR and a new 'Heave to' mode. I guess it would be the equivalent of plane circle mode, try and keep in one spot but not using the GPS.

**Comment by IamPete1** [comment]
@rmackay9  Thanks, once you have finished making small changes I can have a look at putting the current hold mode functionality in a new mode. Or would it be better to remove the hold mode changes from this PR and put them in a new one once this has merged? Unfortunately I will be unable to do any testing in the next few weeks.

**Comment by rmackay9** [comment]
@IamPete1, I've created a [new PR](https://github.com/ArduPilot/ardupilot/pull/9477) with a reduced feature set in the hopes that it is easier for the PR to make it into master.  In total adding sailboat support is a big change so I think breaking it up is a good idea.  I think we should let the other PR get in and then I can help create a new branch which includes the remaining features.

**Comment by rmackay9** [comment]
@IamPete1, I'm working on the 2nd PR now with the tacking feature included.  I'm wondering about the formula to convert the apparent wind speed to an actual wind speed.  The formula is [here in your branch](https://github.com/IamPete1/ardupilot/blob/Sail_2/libraries/AP_WindVane/AP_WindVane.cpp#L438).  I see that it's using the vehicle's forward-back speed but wouldn't it be better to use the vehicle's full 2D velocity vector (that comes from AP::ahrs().get_velocity_NED).  So we could turn the wind speed and direction into a 2D vector, then subtract the vehicle's velocity vector and end up with a final "true" wind speed and direction estimate.

**Comment by IamPete1** [comment]
@rmackay9 Yeah, I think that makes sense.

**Comment by IamPete1** [comment]
This has now all been merged into master, thanks @rmackay9
